### DISCUSSION TITLE
Deprecated `YamuxConfig::set_max_buffer_size`

### DISCUSSION THREAD

**Discussion by MitchTurner** [libp2p/rust-libp2p#5073] [Q&A]
It looks like you're moving to deprecate `set_max_buffer_size` on `YamuxConfig`. What is the recommended replacement for this method? If I don't set this value my peering breaks.

**Comment by mxinden** [comment]
Why does your peering (I am assuming connection) break when you don't set the value?

**Reply by MitchTurner** [reply]
I'm not sure. I have a test that (among other things) tries to publish data for a specific topic via `gossipsub`.  If I remove the max buffer size, the test becomes flaky where it returns the error `InsufficientPeers` when I try to publish. Sometimes it passes.

Before I do debugging to identify what's causing the failed connections, I figured I'd check if you are just moving the contract or abandoning the contract entirely--there isn't any info in the deprecation message:
```rs
#[deprecated(note = "Will be removed with the next breaking release.")]
```

**Reply by thomaseizinger** [reply]
The function is deprecated because the newest yamux version will automatically scale up the buffer based on the latency to the remote peer (see https://en.wikipedia.org/wiki/Bandwidth-delay_product).

Does the issue also occur when you only use `yamux::Config::default()` without any other options? For backwards-compatibility, we only use the new implementation when you don't set any other options. See the source code here: https://github.com/libp2p/rust-libp2p/blob/master/muxers/yamux/src/lib.rs

**Reply by MitchTurner** [reply]
I'll give that a shot! Thanks.

**Reply by MitchTurner** [reply]
Sorry. Just getting to this.

The relevant code is just using `yamux::Config::default()` with the `set_max_buffer_size`:
```
        let multiplex_config = {
            let mplex_config = MplexConfig::default();

            let mut yamux_config = yamux::Config::default();
            #[allow(deprecated)]
            yamux_config.set_max_buffer_size(MAX_RESPONSE_SIZE);
            libp2p::core::upgrade::SelectUpgrade::new(yamux_config, mplex_config)
        };
```

The aforementioned tests fail if I remove `set_max_buffer_size`.

**Reply by thomaseizinger** [reply]
Can you `dbg!` print the config and confirm that it is using `yamux 0.13`?

**Reply by MitchTurner** [reply]
This looks like `0.13` to me:
```
&yamux_config = Config(
    Right(
        Config013(
            Config {
                max_connection_receive_window: Some(
                    1073741824,
                ),
                max_num_streams: 512,
                read_after_close: false,
                split_send_size: 16384,
            },
        ),
    ),
)
```

**Reply by thomaseizinger** [reply]
What does this test exactly do? We have yamux 0.13 working in all our integration tests so I am surprised it fails for you.

**Reply by MitchTurner** [reply]
The test spins up 3 nodes. Node A is connected to Node B, Node B is connected to Node C, but Node C is not connected to Node A.

The test is to see that a message can be published by A and relayed to C.

It fails when A tries to publish the message with the error `InsufficientPeers`.

**Reply by thomaseizinger** [reply]
Wait so your connection does succeed but you can't publish?

**Reply by MitchTurner** [reply]
Node A waits until it has a peer and then tries to publish.

https://github.com/FuelLabs/fuel-core/blob/6df2ff2748db059f9982706c18f07077ef9cc563/crates/services/p2p/src/p2p_service.rs#L1396

The assumption here is that if we get a `PeerInfoUpdated` event and peer is included in the `peer_manager`, then it is included in the `gossipsub`, but maybe there is some race condition in our implementation or perhaps elsewhere that doesn't guarantee this.

I seem to remember debugging something similar to this while upgrading. I'm not sure what the solution was, possilbly reordering some of the behaviors in the `FuelBehaviour`... I had do that for a couple bugs.

Doesn't really explain why the yamux buffer stuff breaks this.

**Reply by thomaseizinger** [reply]
- If you turn up the logs for `libp2p_swarm`, do you see the connection getting established?
- Can you compare logs with and without `set_max_buffer_size` (i.e. version 0.12 and version 0.13)?
- What is your connection timeout? More than `0`?
- Are all peers subscribed to the same topic?
